sklearn/utils/estimator_checks.py

```diff
@@ -1196,6 +1197,64 @@ def check_array_api_input_and_values(
     )
 
 
+def _is_sparse_input_error(e):
+    if not (isinstance(e, TypeError) or isinstance(e, ValueError)):
+        return False
+    return True if re.search("[Ss]parse", str(e)) else False
+
+
+def check_estimator_sparse_tag(name, estimator_orig):
+    """Check that estimator tag related with accepting sparse data is properly set."""
+    if SPARSE_ARRAY_PRESENT:
+        sparse_container = sparse.csr_array
+    else:
+        sparse_container = sparse.csr_matrix
+    estimator = clone(estimator_orig)
+
+    rng = np.random.RandomState(0)
+    n_samples = 15 if name == "SpectralCoclustering" else 40
+    X = rng.uniform(size=(n_samples, 3))
+    X[X < 0.6] = 0
+    y = rng.randint(0, 3, size=n_samples)
+    X = _enforce_estimator_tags_X(estimator, X)
+    y = _enforce_estimator_tags_y(estimator, y)
+    X = sparse_container(X)
+
+    tags = get_tags(estimator)
+    if tags.input_tags.sparse:
+        try:
+            estimator.fit(X, y)  # should pass
+        except Exception as e:
+            raise AssertionError(
+                f"Estimator {name} raised an exception: {e}. The tag "
+                "self.input_tags.sparse might not be consistent with the "
+                "estimator's ability to handle sparse data (i.e. controlled by the "
+                "parameter `accept_sparse` in `validate_data` or `check_array` "
+                f"functions). Got input_tags.sparse={tags.input_tags.sparse}."
+            )
+    else:
+        try:
+            estimator.fit(X, y)  # should fail with appropriate error
+        except Exception as e:
+            if _is_sparse_input_error(e):
+                return
+            else:
+                raise AssertionError(
+                    f"Estimator {name} raised an exception: {e}. "
+                    "The estimator failed when fitted on sparse data in accordance "
+                    f"with its tag self.input_tags.sparse={tags.input_tags.sparse} "
+                    "but didn't raise the appropriate error : error message should "
+                    "state explicitly that sparse input is not supported if this is "
+                    "not the case, e.g. by using check_array(X, accept_sparse=False)."
+                )
```

Comment:
I would find it easier to read without the call to the outside function

```suggestion
        err_msg = (
            f"Estimator {name} raised an exception: {e}. "
            "The estimator failed when fitted on sparse data in accordance "
            f"with its tag self.input_tags.sparse={tags.input_tags.sparse} "
            "but didn't raise the appropriate error : error message should "
            "state explicitly that sparse input is not supported if this is "
            "not the case, e.g. by using check_array(X, accept_sparse=False)."
        )
        try:
            estimator.fit(X, y)  # should fail with appropriate error
        except (ValueError, TypeError) as e:
            if re.search("[Ss]parse", str(e)):
                # Got the right error type and mentioning sparse issue
                return
            raise AssertionError(err_msg) from e
        except:
            raise AssertionError(err_msg) from e
```

Reply:
done in d9f4de3